modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java

```diff
@@ -267,14 +325,45 @@ public GridMetricManager(GridKernalContext ctx) {
      */
     public MetricRegistry registry(String name) {
         return registries.computeIfAbsent(name, n -> {
-            MetricRegistry mreg = new MetricRegistry(name, log);
+            MetricRegistry mreg = new MetricRegistry(name,
+                mname -> readFromMetastorage(metricName(HITRATE_CFG_PREFIX, mname)),
+                mname -> readFromMetastorage(metricName(HISTOGRAM_CFG_PREFIX, mname)),
+                log);
 
             notifyListeners(mreg, metricRegCreationLsnrs, log);
 
             return mreg;
         });
     }
 
+    /**
+     * Reads value from {@link #metastorage}.
+     *
+     * @param key Key.
+     * @param <T> Key type.
+     * @return Value or {@code null} if not found.
+     */
+    private <T extends Serializable> T readFromMetastorage(String key) {
+        metaLock.readLock().lock();
```

Comment:
See my question about this lock above.

Reply:
I want to avoid the following scenario:

0. Metastore not available, yet.
1. T1 creates a metric on node start
2. T1 read config from metastore and got null.
3. T2 execute `onReadyForRead`
4. T2 reads try to configure metric, but it doesn't exists, yet(t1 not finished `addMetric` method execution).
5. T1 finishes.

We metric that not configured with the metastore configuration data.

I fixed it for now, please, see my last commit.

Reply:
Ok. At least now.